### PR TITLE
New generics for slices

### PR DESCRIPTION
**Pull request by michalPyciaTeamwork**

### CODE REVIEW

File: sliceutil/sets.go [hunk 1]
```diff
@@ -57,3 +57,21 @@ func Intersection[T comparable](a, b []T) []T {
 	}
 	return inter
 }
+
+// IntersectionOfMany returns the elements common to all arguments
+func IntersectionOfMany[T comparable](slices ...[]T) []T {
+	if len(slices) == 1 {
+		return slices[0]
+	}
```
Comment (ripexz) [inline sliceutil/sets.go:65]: This is probably not necessary as you have `common = slices[0]` below and with `continue` on first item, it should return immediately after that, albeit it's not bad to have this here 🤔 

On the other hand it might be worth checking `if len(slices) == 0` and then `return nil` as otherwise the line below would panic 😅 Edge case, but if you have some `sliceOfSlices = foo()` followed by `IntersectionOfMany(sliceOfSlices...)` and the result was empty, you'd get an out of bounds panic

Comment (michalPyciaTeamwork) [reply sliceutil/sets.go:65]: @ripexz, great caught!!!
i've rearranged a little bit, now if there are is no input we will return empty slice, this might be easier to handle when using `IntersectionOfMany()` as it will always return same type. We got this covered in tests with `TestIntersectionOfMany():EmptyLists` example

File: sliceutil/sliceutil.go [hunk 2]
```diff
@@ -201,3 +211,17 @@ func Values[T comparable, N any](tt []T, fn func(T) N) []N {
...
+	for _, v := range subset {
+		subsetMap[v] = true
+	}
+	for _, v := range slice {
+		if !subsetMap[v] {
+			return false
+		}
+	}
+	return true
+}
```
Comment (ripexz) [inline sliceutil/sliceutil.go:227]: This func right now is flipped (checking if `slice` is in `subset` rather than the other way round), something like this might work better:

```suggestion
// IsSubset returns if all elements of 'slice' are in 'set'
func IsSubset[T comparable](set, subset []T) bool {
	setMap := make(map[T]bool, len(set))
	for _, v := range set {
		setMap[v] = true
	}
	for _, v := range subset {
		if !setMap[v] {
			return false
		}
	}
	return true
}
```

This is also why your tests are failing as the logic is basically flipped. Also even with the above change the last test would fail, however the expected value of `false` is incorrect as any set is considered a subset of itself.

Also considering this is working with sets, I'd move it to `sets.go` (and tests to `sets_test.go`) rather than the base `sliceutil.go` file.

Lastly, if you'd prefer a bit more readability/reuse over performance, you could simplify it to something like:
```
for _, v := range subset {
	if !Contains(set, v) {
		return false
	}
}
return true
```

File: sliceutil/sliceutil_test.go [hunk 3]
```diff
@@ -307,6 +307,51 @@ func TestItemInSlice_String(t *testing.T) {
...
+
+func TestIsSubset(t *testing.T) {
+	tests := []struct {
+		set      []int
+		subset   []int
+		expected bool
+	}{
+		{[]int{1, 2, 3, 4}, []int{2, 3}, true},
+		{[]int{1, 2, 3, 4}, []int{2, 3, 8}, false},
+		{[]int{1, 2, 3, 4}, []int{1, 2, 3, 4}, false},
```
Comment (ripexz) [inline sliceutil/sliceutil_test.go:342]: As mentioned above, last case should be `true`, I'd also add a couple more tests cases that would be common edge cases (empty slice, nil slice), and potentially move this func to `sets_test.go` 😊

**Review by ripexz — CHANGES_REQUESTED** [review]
Also, while it's not part of the PR, might as well fix the `unused-parameter` lint issues, it's only 4 lines and if you don't see them locally make sure to update your `golangci-lint` 😄
